### PR TITLE
Fix unit test on Python 3.8

### PR DESCRIPTION
**Pull request by bdrung**
Salt fails on Python 3.8:

```
======================================================================
ERROR: unit.grains.test_core (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: unit.grains.test_core
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.8/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "tests/unit/grains/test_core.py", line 37, in <module>
    import salt.grains.core as core
  File "salt/grains/core.py", line 40, in <module>
    from platform import _supported_dists
ImportError: cannot import name '_supported_dists' from 'platform' (/usr/lib/python3.8/platform.py)
```

So only try to import `_supported_dists` from `platform` for Python <=
3.7. Otherwise rely on the external `distro` module to  not need any
special handling.

Addresses parts of #55835

### CODE REVIEW

**Comment by bdrung** [comment]
I pushed more commits to fix all Python 3.8 related unit tests failures.

**Comment by s0undt3ch** [comment]
FYI, I also did some work on this [here](https://github.com/saltstack/salt/tree/features/py38-compat).
I'm not devoting much time to it right now, so feel free to rebase your work on top of that branch, it would be **much** appreciated.

**Comment by bdrung** [comment]
I picked the commits from your branch that I haven't addressed yet.

**Review by cmcmarrow — CHANGES_REQUESTED** [review]
@bdrung thanks for doing this! You’re really close. Just need to get the Windows and Mac test working. Once you have that I think this PR will be good to merge.

**Comment by bdrung** [comment]
The failing Windows and Mac tests are identical with the failings tests of the master branch. So this merge request does not introduce any regression.

**Comment by Ch3LL** [comment]
those tests should be resolved on the head of master so i updated the branch.

**Review by Ch3LL — DISMISSED** [review]

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -306,6 +306,203 @@ repos:
...
+          - --py-version=3.8
+          - --platform=darwin
+          - --include=pkg/osx/req.txt
+          - --include=pkg/osx/req_ext.txt
+          - --include=requirements/base.txt
+          - --include=requirements/zeromq.txt
+          - --include=requirements/pytest.txt
+          - --passthrough-line-from-input=^pyobjc(.*)$
+
+#      - id: pip-tools-compile
```
Comment (waynew) [inline .pre-commit-config.yaml:337]: @s0undt3ch can we add a comment to this part explaining when it needs to be un-commented? I'm okay with pre-staging in this file, but we need something to remind us that it needs to be re-added.

Either that, or just remove it from this PR and open up a separate one that adds it in.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -306,6 +306,203 @@ repos:
...
+        alias: compile-darwin-crypto-py3.8-requirements
+        name: Darwin Py3.8 Crypto Requirements
+        files: ^requirements/(crypto\.txt|static/crypto\.in)$
+        args:
+          - -v
+          - --py-version=3.8
+          - --platform=darwin
+          - --out-prefix=darwin
+
+#      - id: pip-tools-compile
```
Comment (waynew) [inline .pre-commit-config.yaml:388]: @s0undt3ch same here

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -306,6 +306,203 @@ repos:
...
+          - --py-version=3.9
+          - --platform=darwin
+          - --include=pkg/osx/req.txt
+          - --include=pkg/osx/req_ext.txt
+          - --include=requirements/base.txt
+          - --include=requirements/zeromq.txt
+          - --include=requirements/pytest.txt
+          - --passthrough-line-from-input=^pyobjc(.*)$
+
+#      - id: pip-tools-compile
```
Comment (waynew) [inline .pre-commit-config.yaml:436]: @s0undt3ch and here

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -306,6 +306,203 @@ repos:
...
+        alias: compile-darwin-crypto-py3.9-requirements
+        name: Darwin Py3.9 Crypto Requirements
+        files: ^requirements/(crypto\.txt|static/crypto\.in)$
+        args:
+          - -v
+          - --py-version=3.9
+          - --platform=darwin
+          - --out-prefix=darwin
+
+#      - id: pip-tools-compile
```
Comment (waynew) [inline .pre-commit-config.yaml:487]: @s0undt3ch and here

**Comment by s0undt3ch** [comment]
FYI, I pushed a commit that fixes requirements.

**Review by waynew — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (s0undt3ch) [reply .pre-commit-config.yaml:338]: Done

**Comment by bdrung** [comment]
Thanks for adding the fix. I rebased the branch on master and fixed two commits that needed blackening.

**Comment by s0undt3ch** [comment]
Fixed black and pre-commit is passing now

**Review by waynew — DISMISSED** [review]
👍

File: salt/grains/core.py [hunk 5]
```diff
@@ -59,29 +59,34 @@
...
+        "void",
+    )
+
     def linux_distribution(**kwargs):
         with warnings.catch_warnings():
             warnings.simplefilter("ignore")
-            return _deprecated_linux_distribution(**kwargs)
+            return _deprecated_linux_distribution(
+                supported_dists=_supported_dists, **kwargs
+            )
```
Comment (terminalmage) [inline salt/grains/core.py:89]: :+1:

**Comment by bdrung** [comment]
I rebased this patch set on master to resolve the merge commits. This rebase dropped some commits which are already part of the master branch.

**Comment by s0undt3ch** [comment]
It has some lint issues that need fixing now

**Comment by bdrung** [comment]
Fixed the lint issues.

**Review by dwoz — APPROVED** [review]

**Comment by waynew** [comment]
applies to #55835

**Comment by OrangeDog** [comment]
Why did this bother with the try/except on `collections.abc`, when the old location has been deprecated since 3.3, and the minimum supported version for Sodium is 3.5?

**Comment by bdrung** [comment]
Good point. Please open a merge request to drop falling back to the old location.

**Pull request merged**